README.md
# Keccak256 Gem

**Note:** This is a fork of [havenwood/sha3-pure-ruby](https://github.com/havenwood/sha3-pure-ruby) to add support for the (older) Keccak version this hashing algorithm which Ethereum uses.  

> A pure Ruby implementation of Keccak256.

## Installation

Install it from the command line:

```bash
gem install keccak256
```

Or execute `bundle` after adding this line to your app's Gemfile:

```ruby
gem 'keccak256'
```

## Usage Example

```ruby
Digest::Keccak256.new.hexdigest 'cat'
#=> "52763589e772702fa7977a28b3cfb6ca534f0208a2b2d55f7558af664eac478a"
```

## License

Creative Commons

See original license: [havenwood/sha3-pure-ruby](https://github.com/havenwood/sha3-pure-ruby)

ChangeLog.md
### 2.0.0 / 2021-06-04

* Initial Release


test/test_keccak256.rb
require File.expand_path '../../lib/keccak256', __FILE__
require 'minitest/autorun'
require 'minitest/pride'

# http://emn178.github.io/online-tools/sha3_224.html
# http://emn178.github.io/online-tools/sha3_256.html
# http://emn178.github.io/online-tools/sha3_512.html
# http://emn178.github.io/online-tools/sha3_384.html

describe Digest::Keccak256 do
  describe 'Default 256 Keccak padding' do
    it 'should work when empty' do
      empty_string_sha3 = 'c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470'
      assert_equal Digest::Keccak256.new.hexdigest(''), empty_string_sha3
    end

    it 'should work with content' do
      cat_sha3 = '52763589e772702fa7977a28b3cfb6ca534f0208a2b2d55f7558af664eac478a'
      assert_equal Digest::Keccak256.new.hexdigest('cat'), cat_sha3
    end

  end
end
    


test/test_sha3-pure-ruby.rb
require File.expand_path '../../lib/sha3-pure-ruby', __FILE__
require 'minitest/autorun'
require 'minitest/pride'

# http://emn178.github.io/online-tools/sha3_224.html
# http://emn178.github.io/online-tools/sha3_256.html
# http://emn178.github.io/online-tools/sha3_512.html
# http://emn178.github.io/online-tools/sha3_384.html

describe Digest::SHA3 do
  describe 'default 512-bit' do
    it 'should work when empty' do
      empty_string_sha3 = 'a69f73cca23a9ac5c8b567dc185a756e97c982164fe25859e0d1dcc1475c80a615b2123af1f5f94c11e3e9402c3ac558f500199d95b6d3e301758586281dcd26'
      assert_equal Digest::SHA3.new.hexdigest(''), empty_string_sha3
    end
    
    it 'should work with content' do
      cat_sha3 = "fe37dd66fa849ca98684160d542538b22c1edb576271d76b319ded4965d90143a0806fe1edf29b82b8740ec177880769629bdd1a0fb7cb97d7640e60c44833d3"
      assert_equal Digest::SHA3.new.hexdigest('cat'), cat_sha3
    end
  end
  
  describe '384-bit' do
    it 'should work when empty' do
      empty_string_sha3 = '0c63a75b845e4f7d01107d852e4c2485c51a50aaaa94fc61995e71bbee983a2ac3713831264adb47fb6bd1e058d5f004'
      assert_equal Digest::SHA3.new(384).hexdigest(''), empty_string_sha3
    end
    
    it 'should work with content' do
      cat_sha3 = '9bb4adf3004b3ed61f76195321621eac835b6502db486a53b64fdb69c50ee1a8dbb05c950577db70be2bafed59f8891d'
      assert_equal Digest::SHA3.new(384).hexdigest('cat'), cat_sha3
    end
  end
  
  describe '256-bit' do
    it 'should work when empty' do
      empty_string_sha3 = 'a7ffc6f8bf1ed76651c14756a061d662f580ff4de43b49fa82d80a4b80f8434a'
      assert_equal Digest::SHA3.new(256).hexdigest(''), empty_string_sha3
    end
    
    it 'should work with content' do
      cat_sha3 = 'd616607d3e4ba96a74f323cffc5f20a3c78e7cab8ecbdbb03b13fa8ffc9bf644'
      assert_equal Digest::SHA3.new(256).hexdigest('cat'), cat_sha3
    end

  end
  
  describe '224-bit' do
    it 'should work when empty' do
      empty_string_sha3 = '6b4e03423667dbb73b6e15454f0eb1abd4597f9a1b078e3f5b5a6bc7'
      assert_equal Digest::SHA3.new(224).hexdigest(''), empty_string_sha3
    end
    
    it 'should work with content' do
      cat_sha3 = "447c857980c93d613b8bd6897c05bfd0621245139f021aaa6b57830a"
      assert_equal Digest::SHA3.new(224).hexdigest('cat'), cat_sha3
    end
  end

  describe 'Keccak padding' do
    it 'should work when empty' do
      empty_string_sha3 = 'c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470'
      assert_equal Digest::SHA3.new(256, true).hexdigest(''), empty_string_sha3
    end

    it 'should work with content' do
      cat_sha3 = '52763589e772702fa7977a28b3cfb6ca534f0208a2b2d55f7558af664eac478a'
      assert_equal Digest::SHA3.new(256, true).hexdigest('cat'), cat_sha3
    end

  end
end
    


lib/keccak256.rb
require 'digest'

module Digest
  class Keccak256 < Digest::Class
    PILN = [10,  7, 11, 17, 18,  3,  5, 16,
             8, 21, 24,  4, 15, 23, 19, 13,
            12,  2, 20, 14, 22,  9,  6,  1]
            
    ROTC = [ 1,  3,  6, 10, 15, 21, 28, 36,
            45, 55,  2, 14, 27, 41, 56,  8,
            25, 43, 62, 18, 39, 61, 20, 44]
            
    RNDC = [0x0000000000000001, 0x0000000000008082, 0x800000000000808a,
            0x8000000080008000, 0x000000000000808b, 0x0000000080000001,
            0x8000000080008081, 0x8000000000008009, 0x000000000000008a,
            0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
            0x000000008000808b, 0x800000000000008b, 0x8000000000008089,
            0x8000000000008003, 0x8000000000008002, 0x8000000000000080,
            0x000000000000800a, 0x800000008000000a, 0x8000000080008081,
            0x8000000000008080, 0x0000000080000001, 0x8000000080008008]
            
    def initialize
      @size = 256 / 8
      @buffer = ''
    end
    
    def << s
      @buffer << s
      self
    end
    alias update <<
    
    def reset
      @buffer.clear
      self
    end
    
    def finish
      s = Array.new 25, 0
      width = 200 - @size * 2
      padding = "\x01"

      buffer = @buffer
      buffer << padding << "\0" * (width - buffer.size % width)
      buffer[-1] = (buffer[-1].ord | 0x80).chr
      
      0.step buffer.size - 1, width do |j|
        quads = buffer[j, width].unpack 'Q*'
        (width / 8).times do |i|
          s[i] ^= quads[i]
        end
        
        keccak s
      end
      
      s.pack('Q*')[0, @size]
    end
    
    private
    def keccak s
      24.times.each_with_object [] do |round, a|
        # Theta
        5.times do |i|
          a[i] = s[i] ^ s[i + 5] ^ s[i + 10] ^ s[i + 15] ^ s[i + 20]
        end
        
        5.times do |i|
          t = a[(i + 4) % 5] ^ rotate(a[(i + 1) % 5], 1)
          0.step 24, 5 do |j|
            s[j + i] ^= t
          end
        end
        
        # Rho Pi
        t = s[1]
        24.times do |i|
          j = PILN[i]
          a[0] = s[j]
          s[j] = rotate t, ROTC[i]
          t = a[0]
        end
        
        # Chi
        0.step 24, 5 do |j|
          5.times do |i|
            a[i] = s[j + i]
          end
          
          5.times do |i|
            s[j + i] ^= ~a[(i + 1) % 5] & a[(i + 2) % 5]
          end
        end
        
        # Iota
        s[0] ^= RNDC[round]
      end
    end
    
    def rotate x, y
      (x << y | x >> 64 - y) & (1 << 64) - 1
    end
  end
end


lib/sha3-pure-ruby.rb
require 'digest'
require_relative 'sha3-pure-ruby/version'

module Digest
  class SHA3 < Digest::Class    
    PILN = [10,  7, 11, 17, 18,  3,  5, 16,
             8, 21, 24,  4, 15, 23, 19, 13,
            12,  2, 20, 14, 22,  9,  6,  1]
            
    ROTC = [ 1,  3,  6, 10, 15, 21, 28, 36,
            45, 55,  2, 14, 27, 41, 56,  8,
            25, 43, 62, 18, 39, 61, 20, 44]
            
    RNDC = [0x0000000000000001, 0x0000000000008082, 0x800000000000808a,
            0x8000000080008000, 0x000000000000808b, 0x0000000080000001,
            0x8000000080008081, 0x8000000000008009, 0x000000000000008a,
            0x0000000000000088, 0x0000000080008009, 0x000000008000000a,
            0x000000008000808b, 0x800000000000008b, 0x8000000000008089,
            0x8000000000008003, 0x8000000000008002, 0x8000000000000080,
            0x000000000000800a, 0x800000008000000a, 0x8000000080008081,
            0x8000000000008080, 0x0000000080000001, 0x8000000080008008]
            
    def initialize hash_size = 512, keccack = false
      @size = hash_size / 8
      @buffer = ''
      @keccack = keccack
    end
    
    def << s
      @buffer << s
      self
    end
    alias update <<
    
    def reset
      @buffer.clear
      self
    end
    
    def finish
      s = Array.new 25, 0
      width = 200 - @size * 2
      padding = @keccack ? "\x01" :  "\x06"

      buffer = @buffer
      buffer << padding << "\0" * (width - buffer.size % width)
      buffer[-1] = (buffer[-1].ord | 0x80).chr
      
      0.step buffer.size - 1, width do |j|
        quads = buffer[j, width].unpack 'Q*'
        (width / 8).times do |i|
          s[i] ^= quads[i]
        end
        
        keccak s
      end
      
      s.pack('Q*')[0, @size]
    end
    
    private
    def keccak s
      24.times.each_with_object [] do |round, a|
        # Theta
        5.times do |i|
          a[i] = s[i] ^ s[i + 5] ^ s[i + 10] ^ s[i + 15] ^ s[i + 20]
        end
        
        5.times do |i|
          t = a[(i + 4) % 5] ^ rotate(a[(i + 1) % 5], 1)
          0.step 24, 5 do |j|
            s[j + i] ^= t
          end
        end
        
        # Rho Pi
        t = s[1]
        24.times do |i|
          j = PILN[i]
          a[0] = s[j]
          s[j] = rotate t, ROTC[i]
          t = a[0]
        end
        
        # Chi
        0.step 24, 5 do |j|
          5.times do |i|
            a[i] = s[j + i]
          end
          
          5.times do |i|
            s[j + i] ^= ~a[(i + 1) % 5] & a[(i + 2) % 5]
          end
        end
        
        # Iota
        s[0] ^= RNDC[round]
      end
    end
    
    def rotate x, y
      (x << y | x >> 64 - y) & (1 << 64) - 1
    end
  end
end
